02-peer-protocol.md

```diff
@@ -445,6 +921,302 @@ to broadcast the commitment transaction to get his funds back and open a new
+
+Note that `accept_channel`'s `channel_reserve_satoshi` has been omitted.
+Instead, the channel reserve is fixed at 1% of the total channel balance
+(`open_channel2`.`funding_satoshis` + `accept_channel2`.`funding_satoshis`)
+rounded down to the nearest whole satoshi or the `dust_limit_satoshis`,
+whichever is greater.
+
+`zerod_channel_id` for the `accept_channel2` MUST be derived using a zero-d out
+basepoint for the peer's revocation basepoint.  This allows the peer to
+return channel-assignable errors before the *accepter*'s revocation
+basepoint is known.
+
+### Funding Composition
+
+Funding composition for channel establishment v2 makes use of the
+[Interactive Transaction Construction](#interactive-transaction-construction)
+protocol, with the following additional caveats.
+
+#### The `tx_add_input` Message
+
+No additional caveats or requirements.
+
+#### The `tx_add_output` Message
+
+##### Requirements
+
+The sending node:
+  - if is the *opener*:
+    - MUST send at least one `tx_add_output`,  which contains the
+      channel's funding output
+
+##### Rationale
+
+The channel funding output must be added by the *opener*, who pays its fees.
+
+#### The `tx_complete` Message
+
+Upon receipt of consecutive `tx_complete`s, the receiving node:
+  - if is the *accepter*:
+    - MUST fail the negotiation if:
+      - no funding output was received
+      - the value of the funding output is not equal to the sum of
+        `open_channel2`.`funding_satoshis` and `accept_channel2`.
+        `funding_satoshis`
+      - the value of the funding output is less than the `dust_limit`
+ - if it is an RBF attempt:
+    - MUST fail the negotiation if:
+      - the transaction's total fees is less than the last
+        successfully negotiated transaction's fees
+      - the transaction does not share a common input with all previous
+        funding transactions
+
+### The `commitment_signed` Message
+
+This message is exchanged by both peers. It contains the signatures for
+the first commitment transaction.
+
+Rationale and Requirements are the same as listed below,
+for [`commitment_signed`](#commiting-updates-so-far-commitment_signed) with the following additions.
+
+#### Requirements
+
+The sending node:
+  - MUST send zero HTLCs.
+
+The receiving node:
+  - if the message has one or more HTLCs:
+    - MUST fail the negotiation
+  - if it has not already transmitted its `commitment_signed`:
+    - MUST send `commitment_signed`
+  - Otherwise:
+    - MUST send `tx_signatures`
+
+#### Rationale
+
+The first commitment transaction has no HTLCs.
+
+### Sharing funding signatures: `tx_signatures`
+
+After a valid `commitment_signature` has been received
+from the peer and a `commitment_signature` has been sent, a peer:
+  - MUST transmit a [`tx_signatures` message](#the-tx_signatures-message) with their signatures for
+    the funding transaction
+
+#### Requirements
+
+The sending node:
```

Comment:
We should add a requirement to retransmit signatures on reconnection:

```
  - MUST retransmit `tx_signatures` when reconnecting
```

Otherwise we may stumble upon the following dead-lock:

- Alice and Bob have exchanged everything up to `commit_sig`
- Alice goes first and sends her `tx_signatures`
- But Bob disconnects without receiving them
- On reconnection:
  - Alice:
    - is waiting for Bob's `tx_signatures` to be able to publish the funding tx
    - cannot forget the channel because she doesn't know whether Bob has received her signatures or not
  - Bob:
    - has forgotten the channel because he never received Alice's `tx_signatures`
    - cannot prove it to Alice

If peers retransmit their latest `tx_signatures` on reconnection, we get rid of this deadlock.
Does that make sense?

Also, it may be time to squash and rebase :)

Reply:
Thinking about this more, perhaps retransmitting `tx_signatures` is unnecessary complexity.

Alice can handle this situation in the same way she would if Bob intentionally withheld `tx_signatures` (i.e. monitor the blockchain for the pending channel until an input is double spent).  We need to handle this scenario anyway, so retransmitting `tx_signatures` doesn't seem to simplify implementation at all or prevent any griefing attack.

Reply:
> Thinking about this more, perhaps retransmitting tx_signatures is unnecessary complexity.

Why? It really isn't complex at all: just send `tx_signatures` again when reconnecting if you know that channel. At that point you don't have to care about the ordering of signers because you should have forgotten the channel on disconnection if you had not already sent your `tx_signatures` (similarly to how in channel open v1, the accepting node forgets the channel is it gets disconnected before sending `funding_signed`).

It's much easier to explicitly receive your peer's `tx_signatures` (and thus keep a fully signed version of the transaction in your state) than having to monitor the blockchain to extract it IMO.

Reply:
I think I was a bit confused by the Alice/Bob scenario given, since retransmitting doesn't change the outcome there.  Bob forgot the channel.  Alice will resend her `tx_signatures`, and Bob will ignore them or send an error.  So Alice still needs to monitor the blockchain until an input is double spent.

But the reverse scenario is where retransmitting would help.  Bob gets Alice's `tx_signatures` and sends his `tx_signatures`, but Alice disconnects without receiving them.  When Alice restarts, she and Bob will re-exchange signatures, allowing her to finalize the funding tx and broadcast it.

Alternatively, we could go without the re-exchange, and Alice could notice the funding tx was broadcast by Bob.  We need to watch for the funding tx (or any input double-spend) anyway if Bob refuses to send his `tx_signatures`.

> It's much easier to explicitly receive your peer's tx_signatures (and thus keep a fully signed version of the transaction in your state) than having to monitor the blockchain to extract it IMO.

Since we need to handle a missing `tx_signatures` anyway, the code already needs to be implemented.  I think it would be trivial to call it in one new place (on restart).

The second `tx_signatures` is always redundant in the protocol, since the receiver of the first `tx_signatures` can broadcast the funding tx.  So perhaps the root question is: what benefit do we get from sending `tx_signatures` both ways?  And is it worth ensuring we do this in the case of an untimely disconnect?

Reply:
> Since we need to handle a missing tx_signatures anyway, the code already needs to be implemented.

What code are you referring to? We actually don't need to implement monitoring the mempool to get the fully signed transaction: when we don't have the remote signatures we simply don't broadcast the funding transaction (and rely on our peer to do it), but the protocol state machine isn't impacted. Eclair never fetches the funding tx from the mempool when it is missing the remote signatures, and there's no reason to create a dependency on the mempool here.

> The second tx_signatures is always redundant in the protocol, since the receiver of the first tx_signatures can broadcast the funding tx. 

But that's not the end of the story. Consider the case of a mobile wallet and an LSP, where the LSP sends their `tx_signatures` first and doesn't receive the mobile user's `tx_signatures` the first time (because the connection dropped or whatever). The mobile wallet user could regularly go offline for a while, and doesn't necessarily have a very reliable way of broadcasting on-chain txs: since they're offline most of the time, their wallet won't automatically rebroadcast the funding transaction if it gets evicted from the mempool (mobile wallet users simply don't watch the mempool). It is really beneficial for that mobile wallet user to let the LSP regularly rebroadcast the funding transaction to get it confirmed, especially if the mempool is empty at some point while the mobile wallet user is offline. That is only possible if the LSP gets the remote `tx_signatures`, which requires re-sending them.

Reply:
Ack.  The LSP/mobille case is something I wasn't thinking about.  You've convinced me that adding retransmission is worth it.